pep/002-mixin-versioning.md

```diff
@@ -0,0 +1,352 @@
+        digest: def456
+```
+
+When deciding if a mixin resolution can be used, Porter looks at the following:
+
+* The mixin name and url must match.
+* The resolved version must match the version range defined in porter.yaml
+
+If the digest of the binaries do not match the lockfile, the build should return an error.
+
+
+### Mixin Commands
+
+We need to repurpose existing mixin commands now that we have a global mixins cache and a local bundle cache.
+Let's review each mixin command and update its signature with a `--global` flag which indicates that the command should operate against the global cache. This flag defaults to false when the command is executed inside a bundle directory.
+
+
+#### porter mixins search
+
+This command does not operate on the cache so no changes are required.
+
+
+#### porter mixins list
+
+When global is false, list the mixins used by the bundle, taking into account the lockfile.
+
+```
+$ porter mixins list [--global=false]
+Name         Version      Source
+exec         v0.33.1      https://cdn.porter.sh/mixins/atom.xml
+```
+
+The json formatted output includes all the information available from the lockfile.
+
+```json
+[
+ {
+    "name": "exec",
+    "source": "https://cdn.porter.sh/mixins/atom.xml",
+    "url": "https://cdn.porter.sh/mixins/exec",
+    "binaries": {
+      "exec-darwin-amd64": {
+        "digest": "abc123",
+      },
+      "exec-linux-amd64": {
+        "digest": "def456",
+      }
+    },
+    "version": "v0.33.1",
+    "commit": "d433a66",
+    "author": "Porter Authors"
+  }
+]
+```
+
+When global is true, list the mixins in the global cache:
+
+```console
+$ porter mixins list [--global=true]
+Name         Versions
+exec         12
+helm         3
+```
+
+The json formatted output should list all versions in the global cache.
+
+
+#### porter mixins install
+
+When global is false, update the porter.yaml file to declare the specified mixin, install the mixin and cache it in the bundle mixin cache.
+This mode is useful for automating the creation of bundles.
+
+When global is true, install the mixin into the global bundle cache (current behavior).
+This mode is useful for pre-populating the global mixin cache.
+
+
+#### porter mixins upgrade
+
+When global is false, bump the locked version of the mixin per the version range defined in the porter.yaml, and then update the bundle mixin cache.
+This mode is useful for dependabot style CI integrations that automatically keep a bundle up-to-date.
+
+When global is true, upgrade the specified mixin to the latest version in the global bundle cache.
+
+#### porter mixins uninstall
+
+❓ I am not sure if it makes sense to have keep this command? At the bundle level, removing a mixin doesn't seem to have a use case. They could just stop using it. However at the global cache level, what would really be useful is to clear the cache, or removing infrequently used mixin versions. I think cache management like that should be out-of-scope.
```

Comment:
This sounds like the right behavior to have here. Having mixin management at the bundle  level makes it a little confusing as to what's happening. +1 to cache management seems out of scope here.